### PR TITLE
[iOS] Implements infrastructure needed to support CODE_BEEP.

### LINKED ISSUE #257: [iOS] CODE_BEEP should send a vibration or sound
<a href="https://github.com/mcdurdin"></img></a> **Issue by [mcdurdin](https://github.com/mcdurdin)**
_Wednesday Jul 1, 2015 at 22:40 GMT_
_Originally opened as KD-91_

----
Similar to #266 

Since KMW engine handles all codes, it should send a command to vibrate or sound on `CODE_BEEP`

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #257.

This implements a device-appropriate vibration (on supporting devices) whenever a keyboard requires the output of "beep" feedback to the user.

Not all iOS devices support vibrations, though, so a solution still needs to be determined for those.  (Ex:  older-model iPads.)  That said, @darcywong00 says this is also the case for Android, so at least the two will have feature parity in this regard.

### CODE REVIEW

File: ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/keyboard.html [hunk 1]
```diff
@@ -70,6 +71,14 @@
             function getOskHeight() {
                 return oskHeight;
             }
+        
+            function beepKeyboard() {
+                fragmentToggle = (fragmentToggle + 1) % 100;
```
Comment (darcywong00) [inline ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/keyboard.html]: Is it better to move `beepKeyboard()` after the declaration of `fragmentToggle`?  (l. 92)

Could JS handle it as `undefined`?

**Comment by jahorton** [comment]
Glad we tested it, then.  Now to figure out how to get a shorter one.

File: ios/engine/KMEI/KeymanEngine/Classes/Manager.swift [hunk 2]
```diff
@@ -1026,4 +1037,32 @@ public class Manager: NSObject, HTTPDownloadDelegate, UIGestureRecognizerDelegat
...
+
+    if device.isPod {
+      return .none
+    } else if device.isPad {
+      // May not be entirely true, but I can't find any documentation on which
+      // ones DO have it.  Closest thing is https://discussions.apple.com/thread/7415858
+      // that suggests none before Jan 2016 had it, at least.
+      return .none
+    } else if device.isPhone {
+      let basicVibrationModels: [Device] = [.iPhone4, .iPhone4s, .iPhone5, Device.iPhone5s,
```
Comment (jahorton) [inline ios/engine/KMEI/KeymanEngine/Classes/Manager.swift]: Note to self:  this list needs to include the iPhone SE - it doesn't have the Taptic engine:  https://9to5mac.com/2016/03/22/comparison-iphone-se-6-6s-plus-specs/

File: ios/engine/KMEI/KeymanEngine/Classes/KeymanWebViewController.swift [hunk 3]
```diff
@@ -327,6 +331,38 @@ extension KeymanWebViewController: WKScriptMessageHandler {
...
+      //
+      // Ref: https://stackoverflow.com/questions/10570553/how-to-set-iphone-vibrate-length/44495798#44495798
+      // Not usable by older iPhone models.
+      AudioServicesPlaySystemSound(kSystemSoundID_LightVibrate)
+    } else { // if vibrationSupport == .taptic
+      if #available(iOSApplicationExtension 10.0, *) {
+        // Available with iPhone 7 and beyond, we can now produce nicely customized haptic feedback.
+        // We use this style b/c it's short, and in essence it is a minor UI element collision -
+        // a single key with blocked (erroneous) output.
+        let vibrator = UIImpactFeedbackGenerator(style: UIImpactFeedbackStyle.light)
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/Classes/KeymanWebViewController.swift]: This is somewhat too short :) Feels like a legitimate bump. Need something in between!

File: ios/engine/KMEI/KeymanEngine/Classes/KeymanWebViewController.swift [hunk 4]
```diff
@@ -327,6 +331,38 @@ extension KeymanWebViewController: WKScriptMessageHandler {
     // kmw adds w/2 to x.
     return CGRect(x: x - w / 2.0, y: y, width: w, height: h)
   }
+  
+  public func beep(_ keymanWeb: KeymanWebViewController) {
+    let vibrationSupport = Manager.shared.vibrationSupportLevel
+    let kSystemSoundID_LightVibrate: SystemSoundID = 1519
+
+    if vibrationSupport == .none {
+      // TODO:  Find something we can do visually and/or audibly to provide feedback.
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/Classes/KeymanWebViewController.swift]: We have the ability to play sounds ... so we could play a mac "crash" sound ;-) Or something slightly less scary, like a yeti screaming.

**Comment by mcdurdin** [comment]
Okay. After testing on iPhone 7, probably needs to be `.heavy`. It's still not particularly like an error buzz but it's better than 0.4 sec.

For the fallback `kSystemSoundID_MediumVibrate`, `1520` seems better than `1521`. `1521` feels like a double-bump which feels a little weird.

**Pull request merged**